search/collector.go

```diff
@@ -30,3 +30,23 @@ type Collector interface {
 	SetFacetsBuilder(facetsBuilder *FacetsBuilder)
 	FacetResults() FacetResults
 }
+
+// DocumentMatchHandler is the type of document match callback
+// bleve will invoke during the search. 
+// Bleve will indicate the completion of ongoing search, by passing
+// nil value for the document match callback eventually.
+type DocumentMatchHandler func(hit *DocumentMatch) error
+
+type MakeDocumentMatchHandlerKeyType string
+
+var MakeDocumentMatchHandlerKey = MakeDocumentMatchHandlerKeyType(
+	"MakeDocumentMatchHandlerKey")
+
+// MakeDocumentMatchHandler is an optional DocumentMatchHandler 
+// builder function which the applications can pass to bleve.
+// These builder methods gives a DocumentMatchHandler function
+// to bleve, which it will invoke on every document matches in the index.
+// For better memory usage, the DocumentMatchHandler callback implementation
+// should put/give back the hit/DocumentMatch to the ctx.DocumentMatchPool
+// once it is done with the processing of each individual hit.
+type MakeDocumentMatchHandler  func(ctx *SearchContext) (callback DocumentMatchHandler, lookID bool)
```

Comment:
sort of a knee-jerk golanger reaction here... if there's no error in the return vals for stuff like this, it feels like it'll be a delayed future change where you'll be adding a error to the return vals, so might as well add it today :-)